File: neetcode all and misc questions/Graphs/find_the_town_judge.py
class Solution:
    def findJudge(self, n: int, trust: List[List[int]]) -> int:
        """
        adjacency list relationship problem
        for the town judge property, node A points to node B (judge is node B), and node B points to nothing else
        We just have to find the single node that has no outward edges.
        There can only be one node B that points to nothing else
        If the graph is strongly connected (meaning every node points to each other and there are no nodes that have an empty adjacency list), return -1

        Also check that each node points directly at the judge:
        node A points directly at node B, cannot do node A -> middle node -> node B
        
        O(N) Time
        O(N) Space
        """
        adjacency = {}
        for i in range(n):
            adjacency[i+1] = set()
        for a, b in trust:
            adjacency[a].add(b)
        # find the judge
        # if there was more than one node that has an empty list (meaning no neighbors),
        # return False as there can only be one judge
        judge = 0
        for key in adjacency:
            if len(adjacency[key]) == 0:
                # if we already saw a judge, judge would no longer be 0 (assuming judges cannot have the value 0),
                # so this is invalid, as there's more than one judge in the town
                if judge == 0:
                    judge = key
                    break
                return -1
        # check that each node points directly to the judge (excluding the judge himself)
        for key in adjacency:
            if key != judge and judge not in adjacency[key]:
                return -1
        return judge
            

File: node_js_ds_and_algos_implementations/graphs/find_the_town_judge.js
class Solution {
    /**
     * @param {number} n
     * @param {number[][]} trust
     * @return {number}
     */
    findJudge(n, trust) {
        /*
        2/18/2026
        A -> B means A trusts B
        so if every other person trusts the town judge

        A => B <- C

        that means every other element will have an edge in the adjacency list
        except one element, so the answer is just the node that doesn't have any
        incoming edges 

        every node besides the town judge must point at the town judge as well
        */
        let adjacency = {}
        for (let i = 0; i < n; ++i){
            adjacency[i+1] = []
        }
        for (let [a, b] of trust){
            adjacency[a].push(b)
        }
        let potentialJudge;
        for (let person of Object.keys(adjacency)){
            if (adjacency[person].length === 0){
                potentialJudge = parseInt(person)
                break
            }
        }
        if (potentialJudge){
            for (let person of Object.keys(adjacency)){
                person = parseInt(person)
                // if there's no edge to the town judge, this is also an invalid setup,
                // as every other person (besides the judge) must trust the edge
                if (person !== potentialJudge && !adjacency[person].includes(potentialJudge)){
                    return -1
                }
            }
            return potentialJudge
        }
        return -1
    }
}
